File: go/0450M-删除二叉搜索树中的节点.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return root
	}
	if root.Val > key {
		root.Left = deleteNode(root.Left, key)
		return root
	}
	if root.Val < key {
		root.Right = deleteNode(root.Right, key)
		return root
	}
	// 情况 1 : 以下两个 if 已经处理
	if root.Left == nil { // 情况 2 左
		return root.Right
	}
	if root.Right == nil { // 情况 2 右
		return root.Left
	}
	// 情况 3 ： 使用右子树最小元素
	minNode := root.Right
	for minNode.Left != nil {
		minNode = minNode.Left
	}
	root.Val = minNode.Val
	root.Right = deleteNode(root.Right, minNode.Val)
	return root
}

File: cpp/0450M-删除二叉搜索树中的节点.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-23 11:05:54
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root == nullptr) return root;
        if (root->val > key) {  // 向左找
            root->left = deleteNode(root->left, key);
            return root;
        }
        if (root->val < key) {  // 向右找
            root->right = deleteNode(root->right, key);
            return root;
        }
        // 只有一个孩子(包含是叶子，返回nullptr)
        if (root->left == nullptr) return root->right;
        if (root->right == nullptr) return root->left;
        // 两个孩子
        TreeNode* minNode = root->right;  // 右子树最左结点
        while (minNode->left != nullptr) minNode = minNode->left;
        root->val = minNode->val;  // 替代被删除结点
        root->right = deleteNode(root->right, minNode->val);  // 删除替代的结点
        return root;
    }
};